### PR TITLE
Unfinished Clones Have Organ Problems

### PR DESCRIPTION
**Pull request by hbycyl**
This is the patch I promised to write in my comment to #27341 . Here is the algorithm for those interested:

1. For each 10% (rounded up) missing progress, you have a 25% chance to have something in your body fucked up
1a. If the cloner's manipulators have been upgraded, you get a risk-free 10% per tier
2.  For each chance that hits, you lose an external organ. If you accumulated four or more 'hits' you lose an internal organ, besides your brain, instead (and that loss counts for those 4 hits)
2a. External organs will be spit out with you, for 'easy' reattachment

Tested and working, but I would encourage other people to check out the branch of my repo and play around with it to see if it is too punishing or not.

[discussion]
[controversial]
[tweak]
:cl:
 * rscadd: Ejecting clones early can be quite bad for their health!

### CODE REVIEW

**Comment by Dacendeth** [comment]
Are brains considered internal organs? If so this might be a really bad time with bad luck.

**Comment by boy2mantwicethefam** [comment]
It's better to start with external organs and then internal organs

File: code/modules/medical/cloning.dm [hunk 1]
```diff
@@ -429,6 +429,25 @@
...
+			hits++
+		damage_rolls--
+	//apply the damage
+	var/mob/living/carbon/human/H = occupant
+	while(hits>0)
+		if (hits>=4)
+			H.organs_by_name[pick(H.organs_by_name)].droplimb(override = 1, no_explode = 1, spawn_limb = 0, display_message = FALSE)
+			hits -= 4
+		else
+			pick(H.internal_organs).Destroy()
```
Comment (DamianX) [inline code/modules/medical/cloning.dm]:
```suggestion
			qdel(pick(H.internal_organs))
```

**Comment by Kammerjunk** [comment]
all earlier comments apply and are valid

**Comment by PrimeDSS13** [comment]
Like Dacendeth said, having a clone get spat out with no brain/heart would probably be quite the pain. I suppose it's a relatively low chance though, with the amount of organs in a body and it only being 25% of a fuck-up per 10% of progress.

I kind of agree with boy2man as well, even though you can print organs in Medbay it'll be harder to fix people missing multiple organs, as it takes more surgical steps and the patient may have issues in different areas of the body (head, chest, groin, etc).

What if it did a few external organs first, but it spat out the limbs that were fucked up so they're available for re-attachment? (ex: clonebaby is missing both arms, but both arms are spat out of the machine disconnected.)

**Comment by Kammerjunk** [comment]
extra comments also apply: what a terrible idea to delete people's brains (with or without heads) cause a doctor stubbed his toe on the cloner a few times
the only person punished is the patient and the doctor gets to boot people from the round

**Comment by DamianX** [comment]
Definitely don't delete brains.

File: code/modules/medical/cloning.dm (hunk 1 above)
Comment (hbycyl) [reply code/modules/medical/cloning.dm]: /datum/organ/internal overwrites Destroy(), is it still okay to swap it out with qdel?

Comment (DamianX) [reply code/modules/medical/cloning.dm]: qdel ends up calling Destroy, you're not supposed to call Destroy directly even though currently it does almost the same thing.

Comment (hbycyl) [reply code/modules/medical/cloning.dm]: OK, will resolve in next commit to this branch

**Comment by hbycyl** [comment]
Also spits out the detached limbs now

File: code/modules/medical/cloning.dm [hunk 2]
```diff
@@ -429,6 +429,26 @@
...
+	var/hits = 0
+	while(damage_rolls > 0)
+		if(prob(25))//each roll has a 25% chance to give the occupant a bad time
+			hits++
+		damage_rolls--
+	//apply the damage
+	var/mob/living/carbon/human/H = occupant
+	while(hits>0)
+		if (hits>=4)
+			qdel(pick(H.internal_organs - H.internal_organs["brain"]))
```
Comment (DamianX) [inline code/modules/medical/cloning.dm]: `internal_organs` will never contain the string `brain`, you might be thinking of `internal_organs_by_name`

**Review by DamianX — CHANGES_REQUESTED** [review]

File: code/modules/medical/cloning.dm (hunk 2 above)
Comment (hbycyl) [reply code/modules/medical/cloning.dm]: you're right, fixed in latest

**Comment by GreatBigFailure** [comment]
I think external should come before internal like people said, external fucks you up and requires surgery but you're still alive. Most internal organs will kill you fast. Maybe make it a progression thing, like if over 50% progress then it would be further from the important stuff, since of course those would be made first. 
Regardless, great change. Even buffs surgery and a few other less-used medicines/features tangentially.

**Review by DamianX — APPROVED** [review]
I didn't read the rest of the code but he addressed my previous review

**Comment by jknpj** [comment]
external organ damage is piss easy to fix using the monkey panacea

**Comment by Dacendeth** [comment]
Internal too I believe.

**Comment by hbycyl** [comment]
changing the way that works here would be unatomic

**Comment by Kammerjunk** [comment]
and strangely it's something nobody ever raised a huge fuss about

**Comment by tuypo1** [comment]
I think you are severely overestimating how bad a missing limb is compared to a missing organ in terms of difficulty of fixing, it's sure as shit not worse than 4 of them, it's arguably not even worse than one of them

**Comment by tuypo1** [comment]
> external organ damage is piss easy to fix using the monkey panacea

so is organ damage so that's not really relevant here

**Comment by Kammerjunk** [comment]
tried to find what you're replying to but failed. if it's the OP, it's actually the opposite of what point 2 says.

File: code/modules/medical/cloning.dm [hunk 3]
```diff
@@ -429,6 +429,26 @@
...
+			hits++
+		damage_rolls--
+	//apply the damage
+	var/mob/living/carbon/human/H = occupant
+	while(hits>0)
+		if (hits>=4)
+			qdel(pick(H.internal_organs - H.internal_organs_by_name["brain"]))
+			hits -= 4
+		else
+			H.organs_by_name[pick(H.organs_by_name)].droplimb(override = 1, no_explode = 1, spawn_limb = 1, display_message = FALSE)
```
Comment (Kurfursten) [inline code/modules/medical/cloning.dm:447]: this can pick the head or torso, at a minimum the torso needs to not be droplimbed

**Review by Kurfursten — CHANGES_REQUESTED** [review]
do not drop torsos

File: code/modules/medical/cloning.dm (hunk 3 above)
Comment (DamianX) [reply code/modules/medical/cloning.dm:447]: Exclude the groin as well

Comment (hbycyl) [reply code/modules/medical/cloning.dm:447]: Torso and groin do not droplimb, if this pick() lands on them, nothing happens. Relevant snippet from from organ_external.dm:

```
/datum/organ/external/proc/droplimb(formal params)
	if(body_part == (UPPER_TORSO || LOWER_TORSO))
		return
```

Worth adding a comment to note the behavior

Comment (hbycyl) [reply code/modules/medical/cloning.dm:447]: comment added, also changed the PR body to match current version's behavior

**Review by Kurfursten — APPROVED** [review]
It seems like there's an increasingly high chance of no limb loss but if that's intended who am I to argue?

**Pull request merged**